### ISSUE TITLE
RuntimeError when showing dialog from async handler in Cocoa

### ISSUE THREAD

**Issue by samschott** [beeware/toga#1006]
**Describe the bug**
A runtime error "Cannot enter into task while another task is being executed" is raised when showing a dialog from an async callback. This only occurs when there is already a "background task" running, either scheduled manually with `asyncio.create_task` or through toga with `toga.App.add_background_task`. The background task then crashes.

**To Reproduce**

Example code to reproduce, adapted from the handlers example:

<details><summary>Example</summary>
<p>

```Python
import asyncio

import toga
from toga.constants import COLUMN
from toga.style import Pack


class HandlerApp(toga.App):

    async def do_async(self, widget, **kwargs):

        widget.enabled = False
        for i in range(1, 10):
            self.async_label.text = "Iteration {}".format(i)
            if not self.ask_to_continue():
                break

        self.async_label.text = "Ready."
        widget.enabled = True

    def ask_to_continue(self):
        return self.main_window.confirm_dialog('Continue?', 'Press OK to continue.')

    async def do_background_task(self, widget, **kwargs):
        while True:
            self.counter += 1
            self.label.text = "Background: Iteration {}".format(self.counter)
            await asyncio.sleep(1)

    def startup(self):
        self.main_window = toga.MainWindow(title=self.name)

        self.label = toga.Label('Ready.', style=Pack(padding=10))
        self.async_label = toga.Label('Ready.', style=Pack(padding=10))

        self.counter = 0
        self.add_background_task(self.do_background_task)

        btn_style = Pack(flex=1)
        btn_async = toga.Button('Async callback', on_press=self.do_async, style=btn_style)

        box = toga.Box(
            children=[
                self.label,
                btn_async,
                self.async_label,
            ],
            style=Pack(
                flex=1,
                direction=COLUMN,
                padding=10
            )
        )

        self.main_window.content = box
        self.main_window.show()


def main():
    return HandlerApp('Handlers', 'org.beeware.handlers')


if __name__ == '__main__':
    app = main()
    app.main_loop()

```

</p>
</details>

This raises the exception:

```shell
Exception ignored on calling ctypes callback function: <function CFTimerHandle._cf_timer_callback.<locals>.cf_timer_callback at 0x114c4e670>
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/rubicon/objc/eventloop.py", line 155, in cf_timer_callback
    callback(*args)
  File "/usr/local/lib/python3.8/site-packages/rubicon/objc/eventloop.py", line 304, in _callback
    context.run(callback, *args)
RuntimeError: Cannot enter into task <Task pending name='Task-1' coro=<handler_with_cleanup() running at /usr/local/lib/python3.8/site-packages/toga/handlers.py:25> wait_for=<Future finished result=None>> while another task <Task pending name='Task-2' coro=<handler_with_cleanup() running at /usr/local/lib/python3.8/site-packages/toga/handlers.py:25>> is being executed.
```

**Expected behavior**
The background task should pause while the dialog is blocking the event loop and resume afterwards.

**Environment:**
 - Operating System: macOS Catalina, macOS Big Sur
 - Python version: 3.8.5
 - Software versions:
   - Toga: 0.3.0.dev23

**Additional context**
This may be an issue with the Cocoa event loop integration in rubicon.objc.

**Comment by freakboy3742** [comment]
Eep... that's an interesting one. There's probably going to be a similar (but slightly different) problem on iOS; however, in that case, it's because we're trying to synthesize blocking behavior for dialog.

Interestingly, in the process of developing dialogs for Android (#987), @paulproteus has suggested that we make dialog calls asynchronous by default. That's a better match for the Android (and iOS) APIs, and may be a safer API (in the sense that waiting for a dialog won't accidentally cause the background event loop to inadvertently block). The catch will be preserving  (or deciding whether we want to preserve) the "simple" existing behavior for the simple case. However, if we *do* move to an async model for dialogs, that would resolve this problem.

**Comment by SamSchott** [comment]
I agree that async dialogs would be more elegant since we can keep spinning the event loop. This may be difficult to implement with `NSAlert.runModal` which always blocks. One could use `NSAlert.beginSheetModalForWindow:completionHandler:` instead which returns immediately and takes a "completionHandler" as callback to invoke later. In this case, the dialog must be attached to a window, as assumed by the toga API, and will be shown as a sheet.

Unfortunately, app-modal dialogs will no longer be possible this way...

**Comment by SamSchott** [comment]
On the downside, having async dialogs may be a little confusing to Python newcomers. Asyncio is a little confusing by itself, with many different ways of doing things and significant changes during the last releases. In addition, toga's asyncio integration is barely documented.

**Comment by SamSchott** [comment]
Also, maybe this bug can still be fixed with the current synchronous implementation of dialogs. Showing a dialog from an async callback when there are no other background tasks works just fine. Similarly, showing a dialog directly while there is any number of running background tasks also works without issues. There is something about the particular combination of showing a dialog from an async callback while there is another background task which raises an error.

Any idea why this particular combination is problematic?

**Comment by freakboy3742** [comment]
Ok - that makes sense. This is a variant on a ["red code/blue code"](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) problem. Only one async handler can execute at a time; but `showModal()` is an odd function that effectively allows the event loop to continue running. The issue isn't *having* 2 async handlers; it's having the second async handler fire while the first one is being processed. 

We're going to need to be careful with *any* Cocoa API that 'plays nice' with the CoreFoundation event loop. Suspending the Python event loop handling before invoking a CF loop maintaining method would be one approach - although the impact on timed events could be interesting. 

Another approach would be to use primitives to force event loop blocking methods into a separate work task (or thread) so that they don't block the *Python* event loop. This would potentially allow us to make calls to showModal() async for Python purposes (since we can set up a future to contain the showModal() result and await that future).

**Comment by SamSchott** [comment]
Ah, that blog post was a very interesting read! And thanks for explaining what the Python event loop does during calls like `runModal()`.

Ultimately, you’ll have to choose what works best for most platforms, even if this means requiring async functions for an important part of toga functionality...

**Comment by SamSchott** [comment]
I have looked at this in more detail. The issue is caused by asyncio timed callbacks firing even though the event loop should be blocked by a modal dialog. This happens because those timers are added with a runloop mode of `kCFRunLoopDefaultMode`, which includes "default", "modal" and "event tracking" modes. Therefore, those asyncio events are still retrieved while the event loop is restricted to processing modal events.

A quick fix would be to add timer events with the mode `kCFRunLoopDefaultMode` instead, which means that asyncio callbacks cannot fire during modal dialogs. It also means that they won't fire while the event loop is in "event tracking" mode (user interface tracking loops like mouse drags). 

In my quick testing, this works perfectly fine. Do you see any downside to this fix? E.g., do you explicitly want to keep background tasks running during a modal dialog? If yes, a modal dialog is IMO the wrong approach in the first place and async dialogs are better. If you agree, I'll submit a PR against rubicon-objc.

**Comment by freakboy3742** [comment]
That's definitely a neat fix (although I presume you mean using "event tracking" rather than "default" mode as the fix) - but I'm not sure it will work in the general case. 

Consider the case where a the user gets a modal dialog, but they take their time dismissing it (e.g., they walk away from the keyboard and don't respond for 10 minutes). If events are "suspended" while the dialog is displayed, there are two possible outcomes:
 1. The events queue up in a backlog, and once the modal is dismissed, there's a flood of events to process - many of which may be redundant (e.g., in the case of a periodic update timer). It would also be trivial to overflow networking buffers if there is background network events occurring.
 2. Events that occur during the "suspension" are ignored. This avoids the flooding problem from (1) for timer events, but would be a problem if the event is network related, because any received network content would be lost.

The only reason we care about the synchronous invocation is that it's nicer to be able to run "ask_yes_no", get a boolean response, and branch your code as a result. However, I'm starting to think this might be an unachievable goal, and we should move to async alert dialogs (or, at least, rethink the API so they are usable from both a sync and async context).

**Comment by SamSchott** [comment]
There was a typo in the previous post: at the moment, events are added with `kCFRunLoopCommonModes`, which allows them to be processed during a modal loop, and I am proposing to changes this `kCFRunLoopDefaultMode` which is blocked during modal dialogs.

In this case, when asyncio timer events get suspended during a modal dialog, they will indeed queue up in a backlog but won't all be processed at once. Instead, processing will continue at the current pace, as if time had just stopped during the modal dialog. I am not sure about the internals of asyncio but this may be because with a periodic loop, the next event is only scheduled after the previous one has completed. Events are neither dropped nor do they result in flooding. This mirrors the behaviour for a long-running synchronous function call.

Regarding async vs sync dialogs: I don't like keeping the event loop running during what appears to the toga user like a synchronous and blocking call: what looks like a blocking call should behave like a blocking call. If event processing during dialogs is required, those should be explicitly be async and awaited. This can prevent hard to debug issues such as the exception reported here. Also, rubicon-objc should not really bother or know about toga's implementation of dialogs. It should just implements its asyncio integration in the most consistent way. To me, this includes preventing asyncio tasks to be executed during "blocking" Cocoa calls.

But maybe this is just a rant because it was really difficult for me to figure out what exactly was causing the exceptions...

**Comment by massenz** [comment]
>  In addition, toga's asyncio integration is barely documented.

"barely documented" seems even an exaggeration: the ["reference" doc](https://toga.readthedocs.io/en/latest/reference/api/app.html) for `add_background_task(handler)` won't even specify the signature of `handler`, so much so that I have to reverse engineer it :)

(incidentally, AFAICT, it gets invoked with two "app variants"? my main class (a `toga.App`) and a `toga_iOS.app.App`?)

**Comment by kakos-k9** [comment]
I would like to know how you did to call another screen, could you show the code?

**Comment by freakboy3742** [comment]
I believe this has now been resolved by the refactoring performed by #2669 (and previous changes that moved dialogs to be more naturally async.  Please re-open if you can still reproduce the problem.
